lib/app/docker/scanningimageservice.go

```diff
@@ -0,0 +1,145 @@
+	// Example: gravitational/gravity-scan
+	RemoteRepository string
+	// TagPrefix is a string to prepend to each images tag, in order to help identify the source of the image.
+	// Example: 7.0.0 for a release of gravity 7.0.0
+	TagPrefix string
+}
+
+// NewScanningImageService creates an image service that rewrites image paths to a
+// repository used for scanning those images.
+func NewScanningImageService(req RegistryConnectionRequest, conf ScanConfig) (ImageService, error) {
+	base, err := NewImageService(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &scanningImageService{
+		imageService: base.(*imageService),
+		conf:         conf,
+	}, nil
+}
+
+// scanningImageService wraps a normal ImageService and rewrites image paths
+// in a consistent way for pushing all images to a scanning repository
+type scanningImageService struct {
+	*imageService
+	conf ScanConfig
+}
+
+// Sync syncs all images in a local directory, with a remote registry that is capable of scanning and
+// reporting vulnerabilities in those images. All images and tags are written into a designated repository, with the
+// image tag rewritten to help identify the image.
+// Example of gravity 7.0.0:
+//   quay.io/gravitational/debian-tall:0.0.1 -> quay.io/gravitational/gravity-scan:7.0.0_gravitational_debian-tall_0.0.1
+func (r *scanningImageService) Sync(ctx context.Context, dir string, progress utils.Emitter) (installedTags []TagSpec, err error) {
+	if err = r.connect(ctx); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	r.Debugf("Synchronizing local directory %q.", dir)
+	localStore, err := openLocal(dir)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to open local directory %q as local registry", dir)
+	}
+
+	repos, err := ListRepos(ctx, localStore)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to list local repositories in %q", dir)
+	}
+
+	for _, localRepoName := range repos {
+		localRepo, err := localStore.Repository(ctx, localRepoName)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		remoteRepo, err := r.remoteStore.Repository(ctx, r.conf.RemoteRepository)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		localManifests, err := localRepo.Manifests(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		localTags := localRepo.Tags(ctx)
+
+		tags, err := localTags.All(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		for _, tag := range tags {
+			tagSpec := TagSpec{
+				Name: r.conf.RemoteRepository,
+				Version: fmt.Sprintf("%s_%s_%s",
+					r.conf.TagPrefix,
+					strings.ReplaceAll(localRepoName, "/", "_"),
+					tag,
+				),
+			}
+
+			desc, err := localTags.Get(ctx, tag)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			localManifest, err := localManifests.Get(ctx, desc.Digest)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			progress.PrintStep("Pushing image %s:%s -> %s\n", localRepoName, tag, tagSpec)
```

Comment:
I believe the existing implementations will implicitly suffix this with a new line - that is, unless you wanted this to have 2 new line breaks.

Reply:
Yea, I noticed it on the output during testing, but since this was internal code, didn't really bother to track it down, since it's really for internal consumption only. Thanks for pointing it out.